src/lib/supervisor-console.ts

```diff
@@ -0,0 +1,102 @@
+import * as _ from 'lodash';
+import stripAnsi from 'strip-ansi';
+import * as winston from 'winston';
+
+const levels = {
+	error: 0,
+	warn: 1,
+	success: 2,
+	event: 3,
+	api: 4,
+	info: 5,
+	debug: 6,
+};
+
+type logLevel = keyof typeof levels;
+
+const colors: { [key in logLevel]: string | string[] } = {
+	error: 'red',
+	warn: 'yellow',
+	success: 'green',
+	event: 'cyan',
+	info: 'blue',
+	debug: 'magenta',
+	api: ['black', 'bgWhite'],
+};
+
+const maxLevelLength = _(levels)
+	.keys()
+	.map(l => l.length)
+	.max();
+
+const formatter = winston.format.printf(({ level, message }) => {
+	const strippedLevel = stripAnsi(level);
+	return `[${level}]${_.repeat(
+		' ',
+		maxLevelLength! - strippedLevel.length + 1,
+	)}${message}`;
+});
+
+export const winstonLog = (winston.createLogger({
+	format: winston.format.combine(winston.format.colorize(), formatter),
+	transports: [new winston.transports.Console()],
+	// In the future we can reduce this logging level in
+	// certain scenarios, but for now we don't want to ignore
+	// any debugging without a rock solid method of making
+	// sure that debug logs are shown. For example a switch on
+	// the dashboard, a supervisor api call and supervisor
+	// process crash detection
+	level: 'debug',
+	levels,
+	// A bit hacky to get all the correct types for the logger
+	// below, we first cast to unknown so we can do what we
+	// like, and then assign every log level a function (which
+	// is what happens internally in winston)
+}) as unknown) as { [key in logLevel]: (message: string) => void };
+
+winston.addColors(colors);
+
+const messageFormatter = (printFn: (message: string) => void) => {
+	return (...parts: any[]) => {
+		let message = '';
+		let start = '';
+		for (const part of parts) {
```

Comment:
I think the gist of this function is
```ts
parts
	.map((p) => {
		if (p instanceof Error) {
			return p.stack
		}
		return p
	})
	.join(' ')
	.replace('\n','\n  ')
	.split('\n')
	.forEach((p) => {
		printFn(p)
	})
```
right?  Does `printFn` not handle multi-line messages very well?

Reply:
It breaks the formatting a little bit, but it's not the end of the world. This way just makes sure it's consistent. Will change to more functional, the idea was that I'd add more specific handlers for data types, but I didn't get there yet (For example a consistent way of showing docker output etc).

Reply:
You can still add more handlers in the
```ts
	.map((p) => {
		if (p instanceof Error) {
			return p.stack
		}
		return p
	})
```
btw

Reply:
Yeah I know, I was mainly talking about the `switch`, which i will end up adding back in when/if we get more specific handlers.